tasks/hyrax-metadata-updates/tests/test-integration-index.js

```diff
@@ -152,45 +179,252 @@ const event = {
+      body: actual.Body.toString()
+    };
+    const expectedPartial = {
+      etag: outputEtag,
+      body: fs.readFileSync('tests/data/echo10out.xml', 'utf8')
+    };
+
+    t.not(outputEtag, inputEtag);
+    t.like(actualPartial, expectedPartial);
+  } finally {
+    await recursivelyDeleteS3Bucket(t.context.stagingBucket);
+  }
+});
+
+test.serial('hyraxMetadataUpdate eventually finds and updates ECHO10 metadata file with incoming etag', async (t) => {
+  // Set up mock Validation call to CMR
+  nock('https://cmr.earthdata.nasa.gov')
+    .post('/ingest/providers/GES_DISC/validate/granule/MOD11A1.A2017200.h19v04.006.2017201090724.cmr.xml')
+    .reply(200);
+
+  await setupS3(t, false);
+
+  const e = {
+    config: event.config,
+    input: t.context.payload.input
+  };
+
+  try {
+    const metadataFile = t.context.payload.input.granules[0].files.find((f) =>
+      f.type === 'metadata');
+    const bucket = `${metadataFile.bucket}/${metadataFile.fileStagingDir}`;
+    const { ETag: inputEtag } = await getObject(s3(), {
+      Bucket: bucket,
+      Key: metadataFile.name
+    });
+    metadataFile.etag = inputEtag;
+
+    // Upload dummy file to force retries in hyraxMetadataUpdate because ETag is
+    // not initially matched.
+    await promiseS3Upload({
+      Bucket: bucket,
+      Key: metadataFile.name,
+      Body: 'foo'
+    });
+
+    const granulesPromise = hyraxMetadataUpdate(e);
+    await delay(2000).then(promiseS3Upload({
+      Bucket: bucket,
+      Key: metadataFile.name,
+      Body: fs.createReadStream('tests/data/echo10in.xml')
+    }));
+    const { granules } = await granulesPromise;
+
+    const { etag: outputEtag } = granules[0].files.find((f) =>
+      f.type === 'metadata');
+    // Verify the metadata has been updated at the S3 location
+    const actual = await getObject(s3(), {
+      Bucket: `${metadataFile.bucket}/${metadataFile.fileStagingDir}`,
+      Key: metadataFile.name
+    });
+    const actualPartial = {
+      etag: actual.ETag,
+      body: actual.Body.toString()
+    };
+    const expectedPartial = {
+      etag: outputEtag,
+      body: fs.readFileSync('tests/data/echo10out.xml', 'utf8')
+    };
+
+    t.not(outputEtag, inputEtag);
+    t.like(actualPartial, expectedPartial);
+  } finally {
+    await recursivelyDeleteS3Bucket(t.context.stagingBucket);
+  }
+});
+
+test.serial('hyraxMetadataUpdate fails with PreconditionFailure when metadata with incoming etag cannot be found', async (t) => {
+  // Set up mock Validation call to CMR
+  nock('https://cmr.earthdata.nasa.gov')
+    .post('/ingest/providers/GES_DISC/validate/granule/MOD11A1.A2017200.h19v04.006.2017201090724.cmr.xml')
+    .reply(200);
+
+  await setupS3(t, false);
+
+  const e = {
+    config: event.config,
+    input: t.context.payload.input
+  };
+
+  try {
+    const metadataFile = t.context.payload.input.granules[0].files.find((f) =>
+      f.type === 'metadata');
+    metadataFile.etag = randomString();
```

Comment:
This is relying on mutation. Not going to hold the PR up on it, but it's a bit brittle

Reply:
Yes, but these tests are fully isolated in that sense. I would have to either had to modify the fixture files (which I don't want to do since they are used by other tests as well) or rebuild the whole payload, which seems overkill.

Perhaps I'm missing something though. What would you suggest?

Reply:
Given that the payload is coming from the test context, you're right, this is fine as is